ci/src/dependencies/integration/github/github_app.py

```diff
@@ -0,0 +1,90 @@
+import logging
+import os
+import re
+import time
+import typing
+
+import jwt
+import requests
+from model.repository import Repository
+
+GITHUB_APP_CREDENTIALS = {
+    "dfinity": (os.environ.get("GITHUB_APP_CLIENT_ID_DFINITY"), os.environ.get("GITHUB_APP_SIGNING_KEY_DFINITY")),
+    "utopia-icp": (os.environ.get("GITHUB_APP_CLIENT_ID_UTOPIA"), os.environ.get("GITHUB_APP_SIGNING_KEY_UTOPIA")),
+}
+
+if GITHUB_APP_CREDENTIALS["dfinity"][0] is None or GITHUB_APP_CREDENTIALS["dfinity"][1] is None:
+    logging.warning("GITHUB_APP_CREDENTIALS for dfinity are not set, can not clone private repos")
+
+
+class GithubApp:
+    def __init__(self, repos: typing.List[Repository]):
+        self.install_token_by_repo_url: typing.Dict[str, typing.Optional[str]] = {}
+        for repo in repos:
+            if repo.is_private and repo.url.lower().startswith("https://github.com"):
+                self.install_token_by_repo_url[repo.url] = None
+        if len(self.install_token_by_repo_url) > 0:
+            self.__fetch_install_tokens()
+
+    @staticmethod
+    def __gen_jwt(owner: str) -> str:
+        owner = owner.lower()
+        if GITHUB_APP_CREDENTIALS[owner][0] is None or GITHUB_APP_CREDENTIALS[owner][1] is None:
+            raise RuntimeError(f"GITHUB_APP_CREDENTIALS environment variable is not set for {owner}")
+
+        # see https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-json-web-token-jwt-for-a-github-app#example-using-python-to-generate-a-jwt
+        payload = {
+            "iat": int(time.time()),
+            "exp": int(time.time()) + 60,
```

Comment:
Is 60 seconds expiry enough? Since the JWT are instantiated in the beginning and the job may run for a while, do you think we should bump the expiry a bit?